# Введение. Основные понятия теории вероятностей

__Детерминированность__ — это ситуация, в которой одно и то же действие всегда приводит к одному и тому же результату.

Вероятность же имеет дело с неопределённостями.

__Теория вероятностей__ — наука, которая позволяет сделать предположения о более простых вероятностях (об очках на одном кубике) и на их основе математически вывести гораздо более сложные (об очках на нескольких кубиках). К примеру, теория вероятностей помогает ответить на вопрос: «На что ставить, чтобы увеличить шансы на выигрыш?»

__Элементарный исход__ — понятие для любого возможного исхода случайного эксперимента. Например, если обратиться к примеру про количество пользователей на сайте, то все элементарные исходы — это все возможные количества посетителей на сайте.

Разумеется, по результатам эксперимента обычно может получаться много различных результатов (элементарных исходов). Всё множество таких исходов носит называется пространством элементарных исходов и обычно обозначается буквой OMEGA (омега).

## Классическое определение:

Вероятностью случайного события A называется отношение числа n равновероятных элементарных исходов, составляющих событие A, к числу всех возможных элементарных исходов N:

$$P(A) = \frac{n}{N}$$

Элементарные исходы, составляющие событие A, также очень часто называют **исходами, благоприятными или благоприятствующими для события A.**

__Вероятность__ - это число, которое оценивает степень возможности наступления того или иного случайного события.

### Важно!

1. __Вероятность всегда находится в диапазоне от 0 до 1, включительно.__ Это означает, что вероятность какого-либо события не может быть меньше 0 или больше 1. Значение вероятности указывает на степень уверенности, с которой можно ожидать наступление события.

2. __Если вероятность равна 0,__ это значит, что событие __невозможно.__ Оно никогда не произойдет. Например, если подбросить монету и получить на ней обе стороны одновременно, вероятность такого события будет 0.

3. __Если вероятность равна 1,__ это означает, что событие __достоверно.__ Оно обязательно произойдет. Например, если мы говорим о событии "выпал орел или решка" при подбрасывании обычной монеты, то вероятность этого события равна 1, потому что в любом случае монета выпадет с одной из двух сторон.

4. __Значения вероятности между 0 и 1__ отражают __различные уровни вероятности__ наступления события:

- Чем ближе значение вероятности к 1, тем более вероятно, что событие случится.
- Чем ближе значение вероятности к 0, тем менее вероятно, что событие произойдёт.

## Примеры:

Анастасия работает в колл-центре. Сегодня ей необходимо обзвонить 60 клиентов, которые заинтересовались акционным предложением. Она знает, что из 60 человек 23 являются «проблемными» клиентами (надолго затягивают разговор и пишут жалобы на операторов).

Если программа выбирает клиентов из списка случайным образом, то какова вероятность, что день Анастасии начнётся со звонка «хорошему» (то есть не являющемуся «проблемным») клиенту? Ответ округлите до трёх знаков после точки-разделителя.

In [1]:
# Сначала нужно определить количество хороших клиентов, поскольку задача определить вероятность хорошего клиента
good_clients = 60 - 23 # 37

N = 60 # общее количество клиентов
n = good_clients # количество хороших клиентов

A = n / N # вероятность попадания на хорошего клиента
print(f'Вероятность, что Анастасия попадет на хорошего клиента: {round(A, 3)}')

Вероятность, что Анастасия попадет на хорошего клиента: 0.617


Все любимые треки Петра хранятся в плейлисте приложения Spotify, причём 640 из них на английском языке, а 50 — на русском.

Если Пётр устанавливает в настройках случайное воспроизведение треков, то какова вероятность, что он услышит песню на английском языке? Ответ округлите до трёх знаков после точки-разделителя.

In [2]:
# Получить общее количество треков
count_of_tracks = 640 + 50

N = count_of_tracks # общее количество треков
n = 640 # на английском языке

A = n / N # вероятность попадания на трека на английском
print(f'Вероятность, что Петру попадется трек на английском языке: {round(A, 3)}')

Вероятность, что Петру попадется трек на английском языке: 0.928


Пациент участвует в экспериментальном испытании лекарства и услышал, как врачи сказали, что все участники были распределены случайным образом на две группы: 400 человек в тестовой группе (в той, где дают настоящее лекарство) и 500 — в контрольной (в той, где дают плацебо).

Какова вероятность того, что пациент принимает настоящий препарат (то есть находится в тестовой группе)? Ответ округлите до трёх знаков после точки-разделителя.

In [3]:
# Общее количество пациентов
count_of_pacients = 400 + 500
N = count_of_pacients # общее количество пациентов
n = 400 # количество пациентов тестовой группы

A = n / N # вероятность того, что пациент принимает настоящий препарат
print(f'Вероятность, того что пациент принимает настоящий препарат: {round(A, 3)}')

Вероятность, того что пациент принимает настоящий препарат: 0.444


Дополнение события A — это все возможные исходы, кроме A. Их сумма всегда дает 1, так как одно из событий обязательно должно произойти.

Важно понимать, что __у события и дополнения к нему нет общих исходов__, то есть они взаимоисключающие или, как это обычно называют в теории вероятностей, __несовместные__. Также событие и дополнение к нему содержат в сумме абсолютно все исходы из пространства исходов. Из этого следует, что сумма их вероятностей равняется одному (1).

## Примеры:

Вы отправляете посылку почтой. Известно, что посылка дойдёт вовремя с вероятностью 0.34. Также известно, что она никогда не приходит раньше срока. С какой вероятностью посылка придёт позже обозначенного срока?

In [4]:
A_1 = 0.34 # Посылка дойдет вовремя
A_hat = 1.0 - A_1
print(f'Дополнение к событию A_hat: {A_hat}') # Выведет 0.66

Дополнение к событию A_hat: 0.6599999999999999


Для каждого клиента, который берёт кредит в нашем банке, есть три варианта развития событий: вернёт кредит в срок, оформит рефинансирование или выйдет в просрочку.

Известно, что с вероятностью 0.7 клиент вернёт кредит, а с вероятностью 0.1 — выйдет в просрочку. Найдите вероятность того, что клиент оформит рефинансирование кредита.

In [5]:
A_1 = 0.7 # Вернет кредит
A_2 = 0.1 # Выйдет в просрочку
A_refinance = 1.0 - A_1 - A_2 # Оформит рефинансирование
print(A_refinance) # Выведет 0.2

0.20000000000000004


# ПРАВИЛА ТЕОРИИ ВЕРОЯТНОСТЕЙ:

## Правило суммы

Правило суммы — это одно из основных правил в теории вероятностей, которое помогает подсчитать количество благоприятных исходов для двух или более событий. Правило суммы говорит, что если два события не могут произойти одновременно, то общее количество способов, как могут произойти эти события, — это сумма количества способов для каждого события.

### Пример c монеткой
Представь, что ты подбрасываешь монетку. 

Есть два события:

1. Выпал орел.
2. Выпала решка.

Каждое из этих событий может произойти, но одновременно они не могут случиться (или выпал орел, или выпала решка). Если мы хотим узнать общее количество способов, как эти события могут произойти, то просто сложим их:

- Способы для события "орел" — 1 (монетка либо покажет орел, либо нет).
- Способы для события "решка" — 1 (монетка либо покажет решку, либо нет).

Теперь по правилу суммы общее количество способов: 1 + 1 = 2.

### Пример c кубиками
Предположим, что у нас есть два события:

1. На первом кубике выпало четное число (2, 4 или 6).
2. На втором кубике выпало нечетное число (1, 3 или 5).

Мы ищем общее количество способов, как могут произойти эти события. Четные числа на первом кубике: 2, 4, 6 (всего 3 способа). Нечетные числа на втором кубике: 1, 3, 5 (тоже 3 способа).

Применяя правило суммы, общее количество способов будет равно: 3 + 3 = 6.

Значит, общее количество способов — это сумма количества способов для каждого события, если они не могут случиться одновременно.

Когда использовать правило суммы?
- События не пересекаются. Это значит, что одно событие не может произойти одновременно с другим. Например, при подбрасывании монеты не может быть и орел, и решка одновременно.
- Если одно событие происходит, а другое не происходит, то мы можем использовать правило суммы.

## Примеры:

На столе лежат в общей сложности 30 разных книг: 9 — по математике, 10 — по физике и 11 — по химии. Какова вероятность случайно взять книгу, которая не является книгой по математике?

In [6]:
total_books = 30        # Общее количество книг
math_count = 9          # Количество книг по математике
physics_count = 10      # Количество книг по физике
chemistry_count = 11    # Количество книг по химии

non_math_probability = (physics_count + chemistry_count) / total_books
print(f'Вероятность взять книгу не являюшуюся книгой по математике: {non_math_probability}')

Вероятность взять книгу не являюшуюся книгой по математике: 0.7


Вы работаете в компании, организующей доставку различной корреспонденции: письма, рекламные буклеты и журналы. Часть из них отправляется к физическим лицам, а часть — к коммерческим партнёрам. Распределение следующее:

<img src='Images/math_ml_1.png'>

Найдите вероятность, что случайно выбранная посылка...

1) была отправлена физическому лицу. Ответ округлите до трёх знаков после точки-разделителя.

2) является рекламным буклетом или журналом, отправленным коммерческим партнёрам. Ответ округлите до трёх знаков после точки-разделителя.

In [7]:
total = 203 + 97 + 406 + 1021 + 325 + 732 # Общее количество отправлений

physical_probability = (325 + 406 + 203) / total # Вероятность для физ. лица
commercial_probability = (1021 + 97) / total # Вероятность для юр. лица

print(f'Вероятность отправки физическому лицу: {round(physical_probability, 3)}')
print(f'Вероятность отправки юридическому лицу: {round(commercial_probability, 3)}')

Вероятность отправки физическому лицу: 0.335
Вероятность отправки юридическому лицу: 0.402


## Правило умножения

Правило умножения — это еще одно важное правило в теории вероятностей, которое помогает подсчитать количество исходов, когда два события происходят последовательно. Правило умножения говорит, что если одно событие влияет на другое (то есть одно событие должно произойти перед другим), то количество всех возможных исходов можно найти, умножив количество способов для каждого из событий.

### Пример c кубиком
Предположим, ты подбрасываешь два кубика.

- Первый кубик может показать любое из 6 значений (1, 2, 3, 4, 5 или 6).
- Второй кубик тоже может показать любое из 6 значений.

Мы знаем, что результат на первом кубике не зависит от того, какой выпадет результат на втором. Чтобы найти количество всех возможных исходов, применяем правило умножения: 6 × 6 = 36

Здесь:

- 6 — количество возможных исходов на первом кубике.
- 6 — количество возможных исходов на втором кубике.

Итак, всего 36 возможных комбинаций, потому что каждый исход на первом кубике может сочетаться с любым исходом на втором кубике.

### Пример из реальной жизни

Представь, что у тебя есть два рюкзака:

1. В первом рюкзаке 3 футболки: красная, синяя и зеленая.
2. Во втором рюкзаке 2 пары джинсов: черные и синие.

Теперь, если ты хочешь выбрать комплект одежды, то тебе нужно выбрать одну футболку и одну пару джинсов.

Сначала ты выбираешь футболку (есть 3 возможных варианта). После этого выбираешь пару джинсов (есть 2 возможных варианта).

Общее количество комплектов одежды будет равно: 3 × 2 = 6

Это правило работает, потому что выбор футболки не зависит от того, какую пару джинсов ты выберешь, и наоборот. Ты просто умножаешь количество способов для каждого события.

Когда использовать правило умножения?

- Сначала происходит одно событие.
- После этого происходит другое событие.
- Каждый исход первого события может быть связан с любым исходом второго события.

## Примеры:

Футбольная команда выигрывает каждую игру с вероятностью 0.4. Какова приблизительная вероятность того, что они одержат свою первую победу в сезоне в пятой игре? Ответ округлите до двух знаков после точки-разделителя.

Считайте, что результаты в одних играх не влияют на результаты в других играх.

In [8]:
win_probability = 0.4
loose_probability = 1.0 - win_probability

# Мы ищем вероятность события: "Проигрыш, проигрыш, проигрыш, проигрыш, победа"
five_win_probability = loose_probability * loose_probability * loose_probability * loose_probability * win_probability
print(f'Вероятность, что команда выйграет в пятой игре: {round(five_win_probability, 2)}')

Вероятность, что команда выйграет в пятой игре: 0.05


Вероятность того, что в любой день августа в Калининграде пойдёт дождь, составляет 0.09. Какова вероятность того, что дождь пойдёт и 13 августа 2031 г., и 13 августа 2032 г.? Ответ округлите до трёх знаков после точки-разделителя.

In [9]:
raining_probability = 0.09
probability = raining_probability * raining_probability
print(f'Вероятность того, что будет дождь: {round(probability, 3)}')

Вероятность того, что будет дождь: 0.008


## Обобщённое правило суммы

Применение рассмотренного нами правила суммы несколько ограничено, поскольку требует несовместных событий. ?
Что, если события всё же могут быть совместными, то есть происходить одновременно?

Это правило помогает находить вероятность того, что случится хотя бы одно из нескольких событий.

### Основная идея:

Если у тебя есть несколько событий, и они могут произойти по-разному, их вероятности можно сложить, но нужно вычесть вероятность их одновременного наступления (если такие случаи есть).

<img src='Images/math_ml_2.png'>

## Примеры:

В телекоммуникационной компании 34 тысячи клиентов. Из них 14 тысяч используют услугу домашнего интернета, 16 тысяч — услугу спутникового телевидения, а 4 тысячи — обе эти услуги сразу. Ещё 8 тысяч клиентов на данный момент не используют ни одну из перечисленных услуг.

Какова вероятность того, что случайно выбранный клиент использует домашний интернет или спутниковое телевидение или обе эти услуги сразу? Ответ округлите до трёх знаков после точки-разделителя.

In [10]:
total_clients = 34 # Общее количество клиентов (S)
# Данные задачи
P_A = 14 / total_clients # пользуются интернетом
P_B = 16 / total_clients # пользуются спутниковым ТВ
P_AB = 4 / total_clients # пользуются обеими услугами

# Рассчитываем вероятность объединения событий (согласно формуле)
P_A_union_B = P_A + P_B - P_AB

# Выводим результат, округленный до трех знаков после запятой
print(f'Вероятность того, что клиент использует домашний интернет или спутниковое телевидение или обе услуги: {round(P_A_union_B, 3)}')

Вероятность того, что клиент использует домашний интернет или спутниковое телевидение или обе услуги: 0.765


Из стандартной колоды (52 игральных карты) случайным образом выбирается одна карта.

Какова вероятность того, что выбран король или карта червовой масти? Ответ округлите до трёх знаков после точки-разделителя.

In [11]:
total_cards = 52 # Общее количество карт
P_A = 4 / total_cards # Выбран король
P_B = 13 / total_cards # 13 карт составляют черви, будет выбрана одна карта из червей
P_AB = 1 / total_cards # Вероятность, что выбран будет король червей

# Рассчитываем вероятность объединения событий
P_A_union_B = P_A + P_B - P_AB

# Выводим результат, округленный до трех знаков после запятой
print(f'Вероятность того, что будет выбран король червей: {round(P_A_union_B, 3)}')

Вероятность того, что будет выбран король червей: 0.308


# Условная вероятность

Как можно понять из названия, условная вероятность — это вероятность события при некоторых уже известных условиях (вероятность наступления одного события при условии, что уже произошло другое).

Как считается условная вероятность?
Формула:

$$P(A∣B)= \frac{P(A∩B)}{P(B)}$$
​
Где:

- P(A∣B) — вероятность события A, при условии, что уже произошло событие B.
- P(A∩B) — вероятность того, что оба события происходят вместе.
- P(B) — вероятность того, что произошло B (условие).

Например, нам может быть необходимо найти вероятность, что человек вернёт ипотеку, если у него уже есть кредит на автомобиль, или вероятность того, что клиент компании уйдёт к конкурентам, если нам известно, что он отклонял все рекламные предложения в течение предыдущего года.

## Пример со стажером

Представим, что некий стажёр пытается устроиться на работу. С вероятностью 0.06 он успеет вовремя прийти на собеседование и получит работу. С вероятностью 0.02 он в принципе придёт на собеседование вовремя. Необходимо найти вероятность, что стажёр получит работу, если известно, что он уже пришёл на собеседование в нужное время.

Обозначим за N событие, что стажёр пришёл вовремя, а за T — событие, что его взяли на работу. Тогда получаем следующую формулу:


$$ P(T|N) = \frac{P(N∩T)}{P(N)} = \frac{0.06}{0.20} = 0.30$$

Получаем, что стажёр получит работу при условии, что он вовремя пришел на встречу, с вероятностью 0.3.

## Пример с кофе

В конференции участвуют 1000 человек: 300 мужчин и 700 женщин. Известно, что 240 мужчин и 280 женщин пьют кофе. Необходимо найти, с какой вероятностью случайно взятый человек с конференции:

- пьёт кофе, если пол неизвестен;
- пьёт кофе, если это мужчина;
- пьёт кофе, если это женщина.

Для начала опишем всё пространство элементарных исходов. Для простоты введём следующие обозначения:

- OMEGA = {MD, MN, FD, FN}
- M - male
- F - female
- D - drinks
- N - does not drink

### Рассчитаем вероятности для каждого исхода:

- Вероятность, что это мужчина, пьющий кофе:  
  $$ P(MD) = \frac{240}{1000} = 0.24 $$
  
- Вероятность, что это мужчина, не пьющий кофе:  
  $$ P(MN) = \frac{300 - 240}{1000} = 0.06 $$
  
- Вероятность, что это женщина, пьющая кофе:  
  $$ P(FD) = \frac{280}{1000} = 0.28 $$
  
- Вероятность, что это женщина, не пьющая кофе:  
  $$ P(FN) = \frac{700 - 280}{1000} = 0.42 $$

Тогда вероятность того, что человек (мужчина или женщина) пьёт кофе, можем вычислить, используя правило сложения вероятностей:

$$ P(D) = P(\{MD, FD\}) = 0.24 + 0.28 = 0.52 $$

### Рассчитаем условную вероятность того, что человек пьёт кофе, если мы знаем, что это мужчина:

$$ P(D | M) = P(\{MD, FD\} | \{MD, MN\}) = \frac{P(\{MD\})}{P(\{MD, MN\})} = \frac{0.24}{0.3} = 0.8 $$

Аналогично для женщины:

$$ P(D | F) = P(\{MD, FD\} | \{FD, FN\}) = \frac{P(\{FD\})}{P(\{FD, FN\})} = \frac{0.28}{0.7} = 0.4 $$

### Разница между пересечением событий и условной вероятностью

Теперь рассмотрим разницу между вероятностью пересечения событий и условной вероятностью. Найдём вероятность двух похожих, но разных событий:

- Случайно взятый человек с конференции пьёт кофе и является женщиной;
- Случайно взятый человек с конференции пьёт кофе, если уже известно, что это женщина.

Для первого случая используем правило произведения:

$$ P(F \cap D) = P(FD) = 0.28 $$

Для второго случая используем формулу условной вероятности:

$$ P(D | F) = \frac{P(FD)}{P(FD, FN)} = \frac{0.28}{0.7} = 0.4 $$

Можно видеть, что результаты различны, так как это два разных подхода. Будьте внимательны при решении задач, связанных с вероятностями!

## Примеры:

Среди людей, заходящих в магазин, 70 % ничего не покупают. Остальные с равной вероятностью покупают одну-три бутылки воды. Найдите вероятность купить две бутылки воды, если известно, что количество купленных бутылок чётное.

In [12]:
NB = 0.7 # ничего не покупают
PB = 0.3 # остальные покупают 1, 2 или 3 бутылки
P_1 = P_2 = P_3 = PB / 3 # Вероятность купить 1, 2 или 3 бутылки

P_even = NB + P_2 # Вероятность четного количества бутылок (0 или 2)
P_2_given_even = P_2 / P_even # Условная вероятность

print(f'Условная вероятность купить две бутылки при количестве купленных четных: {round(P_2_given_even, 3)}')  # 0.125

Условная вероятность купить две бутылки при количестве купленных четных: 0.125


Аналитик, работающий в спортивном клубе, проанализировал данные и нашёл следующие вероятности:
- P(клиент начал заниматься в январе) - 0.12
- P(клиент занимается в зале более 6 мес) - 0.5
- P(клиент занимается в зале более 6 мес и начал в январе) - 0.024

Найдите вероятность того, что клиент занимается в клубе более шести месяцев, если известно, что он купил абонемент в январе и сразу же приступил к занятиям.

In [13]:
# Заданные вероятности
P_A = 0.12  # Вероятность, что клиент начал заниматься в январе P(A)
P_B = 0.5   # Вероятность, что клиент занимается в зале более 6 месяцев P(B)
P_AB = 0.024  # Совместная вероятность, что клиент начал в январе и занимается более 6 месяцев P(A ∩ B)

# Формула условной вероятности P(B|A) = P(A ∩ B) / P(A)
P_B_given_A = P_AB / P_A

print(f'Вероятность что клиент занимается более шести месяцев при начале в январе: {P_B_given_A}')

Вероятность что клиент занимается более шести месяцев при начале в январе: 0.2


Владелец салона красоты решил подсчитать, какими услугами пользовались его клиенты на прошлой неделе. Получились следующие результаты:

- Окрашивание - 5 %
- Окрашивание и стрижка - 30 %
- Стрижка - 35 %
- Стрижка и макияж - 10 %
- Макияж - 5 %
- Ничего - 15 %

Найдите вероятность того, что случайный клиент захочет сделать стрижку в салоне, если он уже записался на макияж. Ответ округлите до двух знаков после точки-разделителя.

In [14]:
Makeup = 0.05 # Макияж
Haircut = 0.35 # Cтрижка
Makeup_and_Haircut = 0.10 # Макияж и стрижка

# Учесть вероятность, что клиент может прийти либо на макияж, либо на стрижку и макияж
# Полная вероятность макияжа
P_makeup = 0.05 + 0.10 # 0.15

# Нужно найти условную вероятность того, что кл. захочет сделать стрижку если уже записался на макияж
# P(Haircut | Makeup) = P(Makeup and Haircut) / P(Makeup)
P_haircut_given_makeup = Makeup_and_Haircut / P_makeup
print(f'Вероятность стрижки при записи на макияж: {round(P_haircut_given_makeup, 2)}')

Вероятность стрижки при записи на макияж: 0.67


# Независимые события

Независимые события - это такие события, которые никак не влияют друг на друга. Проще говоря, если происходит одно событие, это никак не меняет вероятность наступления другого.

Пример:

1. Бросаем кубик дважды:
- Результат первого броска (например, выпала 6) не влияет на то, что выпадет во второй раз
- Вероятность выпадения любой цифры всегда 1/6, независимо от предыдущих бросков
2. Бросаем монету и кубик:
- Выпадение орла не влияет на то, какая цифра выпадет на кубике

Эти события полностью независимы друг от друга

В жизни:

- То, что вы сегодня проснулись рано, не влияет на вероятность дождя
- Ваша оценка на экзамене не зависит от того, в какой одежде вы пришли

Математически независимость событий проверяется так:

$$P(A и B) = P(A) × P(B)$$

Где:

- P(A) - вероятность события A
- P(B) - вероятность события B
- P(A и B) - вероятность того, что произойдут оба события

Формула проверки:

- Если P(A и B) = P(A) × P(B) - события независимы

- Если P(A и B) ≠ P(A) × P(B) - события зависимы

# Полная вероятность

Теорема полной вероятности помогает найти вероятность события, которое может произойти через разные «сценарии». Например, если вы хотите узнать вероятность дождя, но дождь может идти из разных типов облаков (кучевых, слоистых и т.д.), вы учтёте все варианты и сложите их вклад.

Отсюда вытекает понятие разбиения.

__Разбиение вероятностного пространства__ — это взаимоисключающие и совместно исчерпывающие события. Проще говоря, это события, которые не пересекаются (т. е. не имеют общих исходов), но в объединении дают все возможные исходы.

Как это работает?

__Шаг 1: Разделите задачу на взаимоисключающие условия__
Условия должны:

- Не пересекаться (например, «утро», «день», «вечер»).
- Охватывать все возможности (в сумме их вероятности = 1).

### Пример:

- $B_1$ = «Коробка 1»,
- $B_2$ = «Коробка 2»,
- $B_3$ = «Коробка 3».

__Шаг 2: Посчитайте вероятность события для каждого условия__
Если событие A — «вытащить красный шар», то для каждой коробки нужно знать:

- Вероятность выбрать коробку: $P(B_i)$
- Вероятность вытащить красный шар из этой коробки: $P(A | B_i)$

__Шаг 3: Сложите все вклады__
Формула полной вероятности:

$$P(A)=P(B_1)* P(A∣B_1) + P(B_2) * P(A∣B_2) + ... + P(B_n) * P(A∣B_n)$$

### Пример:

Пусть у нас есть собственная кондитерская, в которой мы продаём пирожные. Каждый клиент обычно покупает равновероятно от 1 до 6 пирожных (получается, что для каждого количества вероятность равна $\frac{1}{6}$), то есть пространство всех элементарных исходов задаётся следующим образом:

OMEGA = {1,2,3,4,5,6}

Рассмотрим следующие события:

A = {1,2}, B = {3}, C = {4,5,6}

- A — клиент купил одно или два пирожных;
- B — клиент купил три пирожных;
- C — клиент купил от четырёх до шести пирожных.

Эти события являются взаимоисключающими, так как у них нет общих исходов!

При этом их объединение даёт все возможные элементарные исходы:

A и B и C = {1,2,3,4,5,6} = OMEGA

Получается, что эти события образуют разбиение вероятностного пространства.

Если рассмотреть ещё один пример в контексте истории с пирожными. Теперь события будут следующими:

A = {1,2}, B = {3,4}, C = {4,5,6}

Такой набор событий __не будет образовывать разбиение__, так как у событий есть общие исходы {4}, а значит, они не являются взаимоисключающими!

# Примеры:

Среди студентов-первокурсников онлайн-магистратуры SkillFactory 30 % устроились на работу в первый месяц обучения, а остальные — позже. Среди тех, кто устроился на работу в первый месяц, 70 % получили повышение в течение года, а среди тех, кто устроился позже, — 50 %.

1. Какие значения удобнее использовать в качестве разбиения в этой задаче?

- Студенты, которые получили повышение в течение года, и студенты, которые не получили повышение в течение года.

2. Какая доля первокурсников SkillFactory устроилась на работу в первый месяц обучения и получила повышение в течение первого года работы?

3. Найдите вероятность того, что случайный первокурсник SkillFactory получил повышение на работе в течение года.

In [15]:
P_early = 0.3 # Студенты которые устроились на работу в течении месяца
P_late = 0.7 # Студенты которые устроились на работу позже

P_raise_given_early = 0.7 # Студенты которые получили повышение в течении года
P_raise_given_late = 0.5 # Студенты которые не получили повышение в течении года

# Правила умножения для студентов которые устроились на работу в первый месяц обучения и получили повышение в течении года
# P(первый месяц | повышение)
P_early_and_raise = P_early * P_raise_given_early
print(f'Доля студентов, которые устроились в первый месяц и получили повышение: {P_early_and_raise}')

# Студенты которые устроились позже и НЕ получили повышения
P_late_raise_given_late = P_late * P_raise_given_late

# Вероятность, что случайный студент (из всех исходов) получил повышение в течение года
# P(повышение | в течение года)
P_year_raise = P_early_and_raise + P_late_raise_given_late
print(f'Вероятность, что случайный студент получил повышение на работе в течение года: {P_year_raise}')

Доля студентов, которые устроились в первый месяц и получили повышение: 0.21
Вероятность, что случайный студент получил повышение на работе в течение года: 0.5599999999999999


Вы едете на деловую встречу в бизнес-центр. Когда на дорогах нет пробок, вероятность не опоздать к назначенному времени составляет для вас 0.9, а когда на улицах пробки, то вероятность успеть вовремя уменьшается до 0.7. Известно, что в том районе, где находится бизнес-центр, пробки бывают в 60 % случаев.

С какой вероятностью вы успеете приехать на деловую встречу без опозданий?

In [16]:
# Данные задачи
P_traffic = 0.6  # Вероятность пробок
P_no_traffic = 0.4  # Вероятность отсутствия пробок

P_on_time_given_no_traffic = 0.9  # Вероятность успеть без пробок
P_on_time_given_traffic = 0.7  # Вероятность успеть с пробками

# Формула полной вероятности
P_on_time = (P_on_time_given_no_traffic * P_no_traffic) + (P_on_time_given_traffic * P_traffic)

print(f'Вероятность успеть на встречу вовремя: {P_on_time:.2f}')

Вероятность успеть на встречу вовремя: 0.78


В одном из отделов банка работает 80 % сотрудников с опытом и 20 % новичков. Вероятность, что опытный сотрудник совершит ошибку при проведении банковской операции, равна 0.01. Вероятность, что такую же ошибку совершит новичок, в 50 раз больше.

Найдите вероятность того, что случайная банковская операция будет произведена без ошибок.

In [17]:
# Данные задачи
P_exp = 0.8 # С опытом
P_no_exp = 0.2 # Без опыта

P_exp_no_mistake = 0.99 # C опытом без ошибок
P_no_exp_no_mistake = 0.5 # Без опыта без ошибок

# Формула полной вероятности
P_no_mistake = (P_exp * P_exp_no_mistake) + (P_no_exp * P_no_exp_no_mistake)

print(f'Вероятность случайной операции без ошибок: {P_no_mistake}')

Вероятность случайной операции без ошибок: 0.892


Вы выяснили, что 2 % клиентов испытали wow-эффект от вашего сервиса и рассказывают о вас буквально всем — из них 70 % приведут новых клиентов. Ещё 20 % клиентов периодически оставляют вам положительные отзывы в соцсетях и на сайте — из них 10 % приведут новых клиентов. Что думают оставшиеся 78 % клиентов (будем называть их «тихонями»), вы не знаете — из них 4 % приведут новых клиентов.

Все ответы округлите до двух знаков после точки-разделителя.

In [18]:
# Данные
P_wow = 0.02
P_new_given_wow = 0.7

P_happy = 0.2
P_new_given_happy = 0.1

P_silent = 0.78
P_new_given_silent = 0.04

# Формула полной вероятности
P_full = (P_wow * P_new_given_wow) + (P_happy * P_new_given_happy) + (P_silent * P_new_given_silent)
print(f'Вероятность, что случайный клиент приведёт нового клиента: {P_full}')

# Какая вероятность, что это сделал клиент-«тихоня»?
P_silent_new_given = (P_silent * P_new_given_silent) / P_full
# Вероятность, что это сделал клиент-wow
P_wow_new_given = (P_wow * P_new_given_wow) / P_full
# Вероятность, что клиент относится к классу "тихоня", если известно, что он НЕ привел нового клиента
P_doesnt_given_wow = 0.3 # Вероятность с коротой он не привел клиента
P_silent_doesnt_new_given = (P_wow * P_doesnt_given_wow) / P_full

print(f'Вероятность, что клиента привел клиент-«тихоня»: {P_silent_new_given:.2f}')
print(f'Вероятность, что клиента привел клиент-"wow": {P_wow_new_given:.2f}')
print(f'Вероятность, что клиент относится к классу "тихоня", если известно, что он НЕ привел нового клиента: {P_silent_doesnt_new_given:.2f}')

Вероятность, что случайный клиент приведёт нового клиента: 0.06520000000000001
Вероятность, что клиента привел клиент-«тихоня»: 0.48
Вероятность, что клиента привел клиент-"wow": 0.21
Вероятность, что клиент относится к классу "тихоня", если известно, что он НЕ привел нового клиента: 0.09


# Теорема Байеса

Является основополагающей теорией, которая помогает __обновлять__ вероятности на основе новой информации.

__Теорема Байеса__ — это основополагающая концепция в области теории вероятностей, которая позволяет рассчитывать точные значения вероятности событий на основе доступной информации. Она особенно полезна в областях, где требуется оценка точности диагноза или предсказаний.

Кроме того, теорема Байеса используется во многих методах машинного обучения, например в спам-фильтрах, чтобы определять, является ли электронное письмо спамом, учитывая слова в этом письме. Кроме того, многие задачи в статистике, например, связанные с интерпретацией медицинских результатов, лучше всего описываются с использованием теоремы Байеса.

### Пример:

Ты детектив. У тебя есть подозреваемый, и ты изначально считаешь, что он виновен с некоторой вероятностью. Но ты нашел новое доказательство - например, отпечатки пальцев. Теперь тебе нужно __пересчитать__ вероятность его вины, учитывая это новое доказательство.

Формула теоремы Байеса:

$$ P(A|B) = \frac{P(B|A) * P(A)}{P(B)}$$

Где:

- P(A|B) - вероятность события A после того, как случилось B (обновление вероятности)
- P(B|A) - вероятность B, если A уже произошло (как часто B случается, когда A истинно)
- P(A) - изначальная вероятность A (априорная вероятность)
- P(B) - общая вероятность B (сумма всех, случаев, когда B может случиться)

### Пример:

1 % клиентов постоянно жалуется на сервис — из них 60 % уйдут. Ещё 10 % клиентов периодически жалуются на сервис — из них 10 % уйдут. Оставшиеся 89 % клиентов никогда не жалуются на сервис — из них 3 % уйдут.

Найдите долю довольных клиентов среди ушедших.

Разобъем клиентов на три группы:

- Жалуются - 0.01
- Иногда - 0.1
- Никогда - 0.89

Затем отметим по группа, кто из них уйдет.

- Жалуются и уйдут - 0.6
- Иногда и уйдут - 0.1
- Никогда и уйдут - 0.03

Сначала получим общую долю всех ушедших клиентов:

P(ушедшие клиенты) = (0.01 * 0.6) + (0.1 * 0.1) + (0.89 * 0.03) = 0.0427

Затем посчитаем долю довольных клиентов (Никогда) среди ушедших:

P(Никогда | ушедшие) = $\frac{(0.89 * 0.03)}{0.0427}$ = 0.63 (63 % ушедших клиентов)

Получается, что больше всего денег мы потеряем из-за тех клиентов, которые были довольны и не жаловались! Это является контринтуитивным предположением, но теория вероятности в том числе нужна для того, чтобы выстраивать стратегию принятия решений на основе чётких доказательств, ведь наше мышление и интуиция часто нас подводят.

Теперь нам необходимо ввести два новых понятия: 

1. Априорные вероятности (безусловные вероятности, то есть они фиксированы и не зависят от вероятностей наступления других событий)

В нашей задаче априорными будут (они даны изначально и не поменяются в задачи):

- P(Жалуются) = 0.01
- P(Иногда) = 0.1
- P(Никогда) = 0.89

2. Апостериорные вероятности (напротив, обусловлены вероятностями наступления каких-то ещё случайных событий)

- P(Никогда | ушедшие) = 0.63 доля довольных клиентов среди ушедших.

### Пример:

Редкое заболевание встречается у 1 % людей. Тест ошибочно показывает наличие болезни у здоровых людей с вероятностью 2 % и ошибочно показывает отсутствие болезни у больных с вероятностью 3 %.

Найдите вероятность, что человек на самом деле болен, если тест на заболевание положителен.

Введём обозначения, чтобы было проще записывать решение:

- S (Sick) - болен
- H (Healthy) - здоровый
- P (Positive) - тест на заболевание положительный
- N (Negative) - тест на заболевание отрицательный

Теперь найдем априорные вероятности, в нашем случае это будут группы:

- P(S) - 0.01 (болен)
- P(H) - 0.99 (здоров)

Также найдём условные вероятности, которые понадобятся нам при решении задачи:

- вероятность того, что тест показывает положительный результат, если человек здоров: 

P(P|H) = 0.99 * 0.02 = 0.2

- вероятность того, что тест показывает отрицательный результат, если человек здоров: 

P(N|H) = 0.99 * 0.98 = 0.98

- вероятность того, что тест показывает положительный результат, если человек болен:

P(P|S) = 0.01 * 0.97 = 0.97

- вероятность того, что тест показывает отрицательный результат, если человек болен:

P(N|S) = 0.01 * 0.03 = 0.3

Пока человек не прошёл медицинский тест, мы считаем его здоровым с вероятностью 99 % и больным — с вероятностью 1 %. Но когда он пройдет тест, у нас появится новая информация и вероятности изменятся!

Давайте найдём полную вероятность, которая складывается из вероятности двух событий:

- P(P|H) P(H) положительный тест, когда человек здоров
- P(P|S) P(S) положительный тест, когда человек болен

P(полная вероятность) = (0.99 * 0.02) + (0.01 * 0.97) = 0.0295 (29,5 %)

Теперь вычислим условную вероятность, пользуясь формулой Байеса:

P(S|P) = $\frac{P(P|S) * P(S)}{P(P)}$ = $\frac{0.97 * 0.01}{0.0295}$ = 0.3288 (0.33 %)

Итак, вероятность того, что человек болен, если получил положительный тест, — всего 0.33. Задумайтесь об этом: этот результат, как и в прошлом примере, может показаться противоречащим интуиции. Это в очередной раз подчёркивает мощность теоремы Байеса: она помогает определять истинные значения вероятностей там, где невозможно оценить их просто по интуиции или из соображений здравого смысла. Теорема Байеса позволяет принимать безошибочные решения на основе вероятностных данных.

# Байесовская статистика

Можно заметить, что теорема Байеса помогает «пересчитывать» априорные вероятности с учётом каких-то дополнительных обстоятельств. Эта её особенность настолько полезна и часто применима, что дала начало новой ветви математической статистики — байесовской статистике.

Байесовская статистика предоставляет математические инструменты для обновления представлений о случайных событиях в свете появления новых данных или свидетельств об этих событиях.

# Наивный байесовский классификатор 

(НБК, англ. Naive Bayes Classifier, NBC) решает задачу классификации объектов по типам. Большим преимуществом этого алгоритма является его простота, как идейная, так и алгоритмическая.

Наивная байесовская классификация — это достаточно простой вероятностный алгоритм, основанный на том, что все признаки модели независимы.

Если, например, мы говорим про задачу классификации спам-писем и обычных писем, то в этом контексте мы считаем, что каждое слово в сообщении не зависит от всех других слов, то есть каждое слово мы учитываем, не обращая внимания на контекст.

Алгоритм классификации выдаёт вероятность того, является письмо спамом или нет, основываясь на наборе слов в письме. Расчёт этой вероятности основан на формуле Байеса, а компоненты формулы рассчитываются на основе частот слов во всём наборе сообщений.

Вероятность того, что письмо является спамом при условии, что в нём есть определённые слова (которые мы обозначили), пропорциональна произведению двух значений:

- вероятности получения спама в целом (по сути, это доля спама в выборке);
- произведения вероятностей, что в письме есть некоторое слово , если письмо является спамом, для всех слов выборки.

Разберёмся подробнее. Для каждого слова в сообщении мы рассчитываем вероятность того, что это слово окажется в спаме. В рамках нашей задачи рассматриваем следующие значения:

- $P(spam)$ — вероятность, что случайно взятое письмо будет спамом (также это доля спам-сообщений в нашем наборе данных);
- $P(w_i|spam)$ — вероятность того, что в сообщении будет определённое слово, если это письмо является спамом.

По той же логике можем определить:

- $P(not spam)$ — доля сообщений, которые не являются спамом;
- $P(w_i|not spam)$ — вероятность того, что в сообщении будет определённое слово, если это письмо не является спамом.

Теперь необходимо понять, как рассчитать вероятности каждого слова. Для этого в алгоритме используется следующая формула:

$$P(w_i|Spam) = \frac{N wi|Spam + \alpha}{N spam + \alpha * N vocabulary}$$

В этой формуле:

- N vocabulary — количество уникальных слов во всём наборе данных;
- N spam — общее количество слов в спам-сообщениях;
- N wi|Spam — количество повторов слова во всех спам-сообщениях;
- $\alpha$ — коэффициент для случаев, когда слово в сообщении отсутствует в нашем наборе данных.

Также эта же формула (но с другими значениями) верна для вероятности того, что слово принадлежит не спам-сообщениям.

# Плюсы и минусы метода

### Плюсы:

- Алгоритм не только прост для понимания и реализации, но также даёт достаточно точные оценки и быстро работает.

- Наивный Байес имеет низкую вычислительную сложность.

- Он может эффективно работать с большим набором данных.

- Его можно использовать с задачами прогнозирования нескольких классов, то есть в задачах мультиклассовой классификации.

- Если выполнено предположение о независимости признаков, то НБК даёт более высокое качество, чем логистическая регрессия и многие другие модели.

### Минусы:

- Предположение о независимых признаках не выполняется на практике практически никогда.

- Если нет обучающего набора данных для какого-то из классов, это приводит к нулевой апостериорной вероятности и модель не может сделать прогноза.

В библиотеке sklearn есть несколько байесовских классификаторов:

- GaussianNB — самый простой вариант, работает с непрерывными признаками;

- MultinomialNB  — работает с категориальными признаками, текстами и несбалансированными выборками;

- ComplementNB — улучшенная версия MultinomialNB, стабильно показывает более высокое качество в задачах классификации текстов;

- BernoulliNB — версия для работы с бинарными признаками;

- CategoricalNB — работает с категориальными признаками, предполагает кодировку данных через OrdinalEncoder.

# Случайная величина и ее характеристики

Случайные величины отличаются от детерминированных (те которые не меняются) наличием неопределённост (могут изменяться взависимости от события).

Случайные величины бывают двух видов — дискретные и непрерывные.

- Дискретные – принимают конечное или счётное количество значений (например, число выпавших очков на игральной кости).

- Непрерывные – могут принимать любое значение из некоторого диапазона (например, рост человека, время ожидания автобуса).

Пример:

- Например, если мы описываем количество клиентов, то это, разумеется, дискретная величина, так как оно может быть выражено только целым неотрицательным числом (не может быть 1.5 клиента или 0.2 клиента).

- К примеру, человеческий рост — это непрерывная величина, так как может принимать бесконечно много значений: 170 см, 170 см и 1 мм, 170 см 1 мм и 0.001 мм и так далее.

### Дискретные случайные величины

Дискретные случайные величины используются в машинном обучении, например в задачах классификации, а также при моделировании распределения слов в тексте для обработки естественного языка.

Представим, что у нас есть распределение выручки в магазине за год. Значение за каждый день нас не особенно интересует (очевидно, что где-то доход будет больше, а где-то — меньше), да и изучать весь набор вероятностей распределения, чтобы разобраться в тенденциях и спрогнозировать прибыль на будущее, может быть довольно долго и трудозатратно. Нам хотелось бы описать распределение всех значений кратко и ёмко. Например, «В среднем ожидаемая выручка в этом магазине составляет 50 тысяч рублей в день, причём в разные дни колебания значений не очень большие, в районе 10 тысяч».  Для того чтобы это сделать, нам понадобится научиться искать два значения: математическое ожидание (оно будет помогать найти центр распределения) и дисперсию (с её помощью мы будем оценивать разброс значений).

__Математическое ожидание случайной величины__ — это взвешенное среднее значений этой величины с учётом вероятностей. Чаще всего оно обозначается буквой E (от англ. Expectation), ещё можно общепринятое обозначение — $\mu$ (греческая буква «мю»). В русскоязычной литературе принято обозначать математическое ожидание буквой M.

### 1. Математическое ожидание:
$$
E(X) = \sum X_i P(X_i)
$$
Где:
- \( X_i \) — возможные значения случайной величины,
- \( P(X_i) \) — вероятности, с которыми эти значения принимаются.

### 2. Математическое ожидание \( X^2 \):
$$
E(X^2) = \sum X_i^2 P(X_i)
$$
Где:
- \( X_i^2 \) — квадраты возможных значений случайной величины.

Чтобы корректно описывать ряд данных, также нужно уметь оценивать и разброс (дисперсию).

### 3. Дисперсия:
$$
V(X) = E(X^2) - (E(X))^2
$$
Где:
- \( E(X^2) \) — математическое ожидание \( X^2 \),
- \( (E(X))^2 \) — квадрат математического ожидания \( X \).

Если внимательно посмотреть на формулу, можно увидеть, что для вычисления дисперсии приходится возводить значения в квадрат. Это очень неудобно для интерпретации, так как непонятно, как содержательно оценивать разброс в 100 рублей в квадрате или 15 метров в четвёртой степени. Поэтому на практике обычно используют не саму дисперсию, а корень из неё — стандартное отклонение. 

### 4. Стандартное отклонение:
$$
\sigma(X) = \sqrt{V(X)} = \sqrt{E(X^2) - (E(X))^2}
$$
Где:
- \( V(X) \) — дисперсия случайной величины \( X \).

## Полный пример:

<img src='Images/math_ml_3.png'>

# Равномерное распределение

Равномерное распределение — это распределение вероятностей, в качестве значений которого могут выступать любые целые числа от 1 до N, а вероятности их появления одинаковы.

Равномерное распределение применяется там, где нужны равновероятные модели.

Например, __равномерное распределение__ пригодится, если для розыгрыша призов необходимо реализовать генератор случайных чисел, который будет с одинаковыми вероятностями выбирать натуральные числа в заданном интервале.

Для равномерного распределения уже известны: 

- __матиматическое ожидание:__

$$EX = \frac{N + 1}{2}$$

- __стандартное отклонение:__

$$ \sigma_x = \sqrt{\frac{N^2 - 1}{12}}$$


### Пример:

<img src='Images/math_ml_4.png'>

Для матиматического ожидания:

$$EX = \frac{N + 1}{2} = \frac{6 + 1}{2} = 3.5$$

Далее для стандартного отклонения:

$$\sigma_x = \sqrt{\frac{N^2 - 1}{12}} = \sqrt{\frac{6^2 - 1}{12}} = 1.7$$

Получаем, что ожидаемое количество купленных пирожных — 3.5 со стандартным отклонением в 1.7.

# Распределение Бернулли

__Распределение Бернулли,__ по сути, моделирует однократное подбрасывание «фальшивой» монеты. Это распределение вероятностей случайной величины, принимающей только два значения: 1 («успех») и 0 («неудача») с вероятностями p и 1 - p соответственно. Таким образом, распределение Бернулли __описывает события, имеющие ровно два исхода.__ Такие события повсеместно встречаются в реальной жизни (выиграет команда чемпионат или нет, сдаст студент экзамен или провалит его, распознает алгоритм объект на фото или нет).

Основной параметр, определяющий это распределение, — p (вероятность «успеха»).

Для этого распределения также известны:

- математическое ожидание:

$$EX = 1 * p + 0 * (1 - p) = p$$

- стандартное отклонение:

$$\sigma_x = \sqrt{p(1 - p)}$$

Вы стреляете по мишени в тире. Вероятность попасть составляет 0.6. В случае успеха вы выиграете плюшевого медведя стоимостью 3000 рублей, а в случае промаха — не выиграете ничего. У вас есть только один выстрел.

Какой должна быть стоимость билета, чтобы игра была честной, то есть чтобы цена билета равнялась ожидаемому выигрышу, а значит, не уводила в убыток продавца или вас при большом количестве выстрелов?

Тогда:

Пусть случайная величина X принимает значение 1, если вы попали, и 0, если промахнулись. Тогда ожидаемое число попаданий равно:

1 * 0.6 + 0 * 0.4 = 0.6

Ожидаемый выигрыш:

0.6 * 3000 + 0.4 * 0 = 1800

Получаем, что «справедливая» стоимость билета должна быть равна 1800 рублей.

# Биноминальное распределение

Повторение нескольких независимых испытаний Бернулли называется __процессом Бернулли.__ Результаты процесса Бернулли в свою очередь будут следовать __биномиальному распределению.__

__Биномиальным__ называют распределение, при котором возможны только два исхода (успех или неудача, выигрыш или проигрыш) и вероятность успеха и неудачи одинакова для всех испытаний. Однако исходы не обязательно должны быть равновероятными, и каждое испытание не зависит от других.

Параметры биномиального распределения — n и p, где n — общее количество испытаний, а p — вероятность успеха в каждом испытании.

<img src='Images/math_ml_5.png'>

- математическое ожидание:

$$EX = np$$

- стандартное отклонение:

$$\sigma_x = \sqrt{n * p * (1 - p)}$$

Сравним две рекламных кампании: в одной конверсия составляет 2 %, но баннер увидит 100 000 пользователей, а в другой конверсия 40 %, но охват — всего 5 000 пользователей.

У нас есть следующая величина для количества пользователей, кликнувших по рекламе в первом случае:

X ~ Binominal(nx = 100_000, px = 0.02)

Случайная величина, отражающая количество пользователей, перешедших по рекламе, для второй рекламной кампании:

Y ~ Binominal(ny = 5000, py = 0.4)

Математическое ожидание и разброс для первой кампании:

$$EX = nxpx = 100000 * 0.02 = 2000$$

$$\sigma_x = \sqrt{nxpx(1 - px)} = \sqrt{100000 * 0,02 * 0.98} = 44$$
человека

Математическое ожидание и разброс для второй кампании:

$$EX = nxpx = 5000 * 0.4 = 2000$$

$$\sigma_x = \sqrt{nypy(1 - py)} = \sqrt{5000 * 0,4 * 0.6} = 34$$
человека

Получаем, что ожидаемое количество кликов в обеих кампаниях одинаковое, так что более выгодным вариантом было бы выбрать менее затратную с финансовой точки зрения.

Также можно сказать, что в первой стратегии разброс значений больше, а значит она менее стабильная.

Производитель гаджетов знает, что 20 % производимых им товаров — бракованные.

Если он производит десять изделий в день, какова вероятность того, что не более двух из них бракованные?

<img src='Images/math_ml_6.png'>

# Распределение Пуассона

Распределение Пуассона — это дискретное распределение вероятностей числа событий, происходящих в данный период времени, с учётом среднего количества раз, когда событие происходит за этот период времени.

Допустим, у нас есть ресторан быстрого питания, в который приходят  в среднем Три посетителя в минуту. Однако это всего лишь средний показатель — фактический показатель может варьироваться.

Распределение Пуассона можно использовать для анализа вероятности различных событий, касающихся того, сколько клиентов придёт в ресторан. С его помощью можно рассчитать вероятность затишья (когда в течение какого-то периода времени не будет ни одного клиента), а также вероятность всплеска активности (когда в ресторан за минуту придут пять и более клиентов). Эта информация, в свою очередь, может помочь менеджеру спланировать рабочую нагрузку и график сотрудников, а также план по закупкам продуктов.

Кроме использования в планировании нагрузки на персонал и производство, распределение Пуассона также используется в биологии (обнаружение мутаций), финансах (прогнозирование количества заявлений о банкротстве или просрочек кредитов), информационной безопасности (предсказание количества вирусных атак) и любой другой ситуации, в которой события зависят от времени.

В распределении Пуассона значение случайной величины может быть любым неотрицательным числом.  Случайная величина будет обладать следующими характеристиками:

- математическое распределение:

$$EX = \lambda$$

- стандартное отклонение:

$$\sigma_x = \sqrt{\lambda}$$

где $\lambda$ — ожидаемое число событий за период времени.

<img src='Images/math_ml_7.png'>

Колл-центр получает в среднем 4.5 звонка за каждые пять минут. Каждый оператор может обработать один из этих вызовов в течение пяти минут. Если вызов получен, но оператор не может его принять, то вызывающий абонент будет переведён в режим ожидания ответа.

Если вызовы следуют распределению Пуассона, какое минимальное количество операторов необходимо колл-центру, чтобы вероятность попадания вызова на удержание была не более 0.1?

Рассчитаем вероятности для всех возможных количеств операторов, пока не получим вероятность того, что вызов находится на удержании, которая будет меньше 10 %:

<img src='Images/math_ml_8.png'>

В качестве параметра $\lambda$ мы берём среднее количество звонков, так как это наше ожидаемое число событий, а в качестве $k$ — количество операторов, т. к. в данный период времени мы сможем принять столько звонков (один оператор принимает один звонок).

Получаем, что колл-центру необходимо нанять на работу хотя бы семь операторов.

#  Непрерывные распределения

Представьте, что вы анализируете процессы всемирного потепления, и вам необходимо фиксировать точные значения температуры в разных точках планеты. Разумеется, у вас есть сверхточные инструменты, поэтому температура может принимать огромный спектр значений: не просто 23 или 24 градуса, а 23.14, 23.5 или даже 23.4582 градуса. В таком случае даже очень большая и подробная таблица не сможет описать все возможные значения, вероятность каждого конкретного значения размывается, и работать с моделью становится сложно.

Для таких ситуаций существует __непрерывная модель.__ В ней распределение вероятностей задаётся не таблицей, а с помощью функции __плотности вероятности.__

<img src='Images/math_ml_9.png'>

Как и в дискретном случае, мы рассмотрим основные непрерывные распределения:

- равномерное;
- нормальное;
- экспоненциальное.

# Равномерное распределение

Непрерывная случайная величина X распределяется равномерно, если вероятность попадания X в интервал пропорциональна длине этого интервала. То есть, к примеру, если мы рассматриваем все числа от 1 до 100 и ищем вероятность попадания в интервал от 1 до 10, то вероятность равна 0.1, а для интервала от 1 до 30 — 0.3.

__Равномерную непрерывную случайную величину__ можно использовать для генерации случайных чисел, если подходят любые вещественные значения. Генерация случайных чисел (по сути, генерация «шума»), в свою очередь, используется для того, чтобы получать из маленькой выборки __бόльшую__ (это называется __аугментацией__): если вы умеете генерировать непрерывные распределения (в том числе равномерное), то можете искусственно увеличить объём ваших данных и повысить качество модели.

Случайная величина, которая распределена по непрерывному равномерному закону, может принимать в качестве значения любое вещественное число в заданном отрезке от a до b и имеет следующее: 

- математическое ожидание:

$$ EX = \frac{a + b}{2}$$

- среднеквадратичное отклонение:

$$\sigma_x = \frac{b - a}{\sqrt{12}}$$

Предположим, что распределение определено на отрезке от 4 до 8 и нам нужно найти для него математическое ожидание и стандартное отклонение.

Для начала найдём математическое ожидание. В качестве a подставляем 4, а вместо b — 8. Получаем (4 + 8) / 2 = 6. Это значит, что среднее (взвешенное по вероятностям возможных значений) значение нашей случайной величины — это 6.

Также вычислим стандартное отклонение. Мы уже выяснили, что a = 4, b = 8. Значит, осталось только выполнить расчёты:

$$\frac{(8 - 4)}{\sqrt{12}} = 1.16$$

Ранее мы говорили, что для того, чтобы найти вероятность попадания в какой-то интервал, необходимо использовать плотность распределения. Здесь она постоянна и имеет следующий вид:

- формула протность вероятности:

$$
f(x) =
\begin{cases} 
0, \text{ при } x < a, x > b \\ 
\frac{1}{b-a}, \text{ при } a \leq x \leq b 
\end{cases}
$$

Это значит, что для любых точек вне рассматриваемого интервала вероятность попасть в него равна нулю, а для точек внутри интервала вероятность равна $\frac{1}{b - a}.$

Для примера, рассмотренного ранее, вероятность попасть в единичный интервал из промежутка от 4 до 8 равна:

$$
\frac{1}{8 - 4} = 0.25.
$$

# Нормальное распределение

Нормальное распределение (его ещё называют гауссовским или распределением Гаусса) представляет собой распределение вероятностей, обычно используемое для моделирования таких явлений, как физические характеристики (например, рост, вес и т.д.) или результаты тестирований.

Если, например $X = N(4, 5^2)$, то

- $\mu$ = 4 математическое ожидание
- $\sigma = 5^2$ (25) дисперсия

- математическое ожидание:

$$EX = 4 (\mu)$$ 

- стандартное отклонение:

$$\sigma_x = \sqrt{25}$$

Кроме того, что по нормальному закону распределены многие величины, нормальное распределение очень часто используется в статистике и машинном обучении, чтобы проверить корректность построенной в модели. К примеру, из задачи регрессии, если в данных есть остатки (или их ещё называют отклонениями или ошибками) — разница между реальными и предсказанными значениями. Оказывается, чтобы убедиться, что модель корректная, необходимо собрать все эти остатки и проверить, распределены ли они по нормальному закону. Это можно сделать и визуально — форма нормального распределения легко узнаваема и выглядит следующим образом:

<img src='Images/math_ml_10.png'>

Видно, что у данного распределения очень большая плотность около середины и намного меньше — у краёв. Если бы мы построили график для остатков в модели линейной регрессии и получили такой результат, мы могли бы сделать вывод, что с точки зрения этого критерия модель построена корректно.

<img src='Images/math_ml_11.png'>

Среди всех нормальных распределений выделяют одно особенное — __стандартное нормальное.__ У него нулевое среднее и единичная дисперсия.

График его плотности изображён ниже:

<img src='Images/math_ml_12.png'>

Для нас представляет интерес переход от произвольного нормального распределения к стандартному.

Если у нас есть нормальное распределение, то, чтобы свести его стандартному, необходимо из каждого значения вычесть среднее и разделить на стандартное отклонение:

$$z = \frac{X - \mu}{\sigma}$$

Чем это может быть полезно? Допустим, нам нужно сравнить результаты тестирования двух групп студентов, которые сдавали разные экзамены на знание английского языка, TOEFL и IELTS. Первый экзамен оценивается по шкале от 0 до 120, а второй — от 0 до 9. Как сравнить результаты, измеренные в разных единицах, между собой? Здесь нам и помогает стандартизированное распределение. Если мы переведём оценки обеих групп студентов в так называемые z-оценки, то сможем сравнить результаты между группами. 

# Экспоненциальное распределение

Экспоненциальное распределение — это распределение вероятностей времени между событиями в процессе Пуассона.

В реальной жизни интервалы времени между событиями обычно разные. Например, если мы анализируем количество поступающих на рабочую почту писем, то оно может быть разным в зависимости от дня и времени суток. Но если мы возьмём небольшой временной интервал, в течение которого скорость примерно постоянная, например с 10:30 до 14:30 в понедельник, экспоненциальное распределение можно использовать как хорошую приблизительную модель для времени до поступления следующего письма.

Значениями случайной величины, распределённой по экспоненциальному закону, являются любые вещественные неотрицательные числа.

Плотность экспоненциального распределения задаётся следующей формулой:

$$f(x) = 
\begin{cases}
0, \text{ при } x < 0 \\ 
\lambda_e^-\lambda_x, \text{ при } x >= 0 
\end{cases}$$

В качестве параметра $\lambda$ здесь берётся ожидаемое число событий за единицу времени.

Математическое ожидание для этого распределения вычисляется по следующей формуле:

$$EX = 1/\lambda$$

, где 1 - среднее значение распределения, для экспоненциального распределения это также и математическое ожидание.

Стандартное отклонение можно найти ровно так же:

$$\sigma_x = 1/\lambda$$

Вероятность того, что значение случайной величины будет не больше x, равна:

$$1 - e^-\lambda_x$$

Применим для решения задачи формулу, которую рассмотрели выше:

$$P(X < 10) = 1 - e^-\lambda(10) = 0.4865$$

В качестве параметра распределения здесь используется $\lambda = 1/15$.

Получаем, что с вероятностью 0.49 случайно выбранный путешественник купит авиабилет менее чем за 10 дней до рейса.